AX: CSS content alt text is not announced by VoiceOver

twilco opened the pull request:
#### 89606797423e2fcdb8fa714756b0c9233500cfc3
<pre>
AX: CSS content alt text is not announced by VoiceOver
<a href="https://bugs.webkit.org/show_bug.cgi?id=292653">https://bugs.webkit.org/show_bug.cgi?id=292653</a>
<a href="https://rdar.apple.com/150830550">rdar://150830550</a>

Reviewed by Joshua Hoffman.

This commit incorporates usage of alt text from CSS `content` into the various existing places we get alt text, e.g.
from the alt attribute.

* LayoutTests/accessibility/css-content-alt-text-expected.txt: Added.
* LayoutTests/accessibility/css-content-alt-text.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
(WebCore::AccessibilityImageMapLink::description const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::usesAltForTextComputation const):
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::description const):
(WebCore::AccessibilityNodeObject::usesAltTagForTextComputation const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::altTextFromAttributeOrStyle const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::objectInclusionFromAltText):
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::updateAltText):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::altFromContent const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/297848@main">https://commits.webkit.org/297848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec902ac26be3206b9c7c19374c84ca61767eaba5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23314 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119308 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63699 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41399 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86082 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/37587 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116048 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66398 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26018 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19868 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63063 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96129 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19941 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122526 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29965 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94929 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97956 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94670 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24161 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39817 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17622 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36304 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40065 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45564 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39706 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41443 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on LayoutTests/accessibility/css-content-alt-text.html:
```diff
@@ -0,0 +1,57 @@
+}
+
+.alt-text-2 {
+    content: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='24' height='24'><circle cx='12' cy='12' r='12' fill='purple'/></svg>") / "alt-text-2";
```
Chromium has a bunch of test cases here:

https://source.chromium.org/chromium/chromium/src/+/main:content/test/data/accessibility/css/alt-text.html

Could you check those and see if they catch any new edge cases, or if there are any that we should also cover in testing.

twilco replied:
Reviewed these testcases, comparing behavior against Chromium, and filed two bugs to follow-up on:

https://bugs.webkit.org/show_bug.cgi?id=296451 (AX: VoiceOver doesn't read alt-text of CSS content in place of unicode string value)
https://bugs.webkit.org/show_bug.cgi?id=296451 (AX: VoiceOver reads "alt, group" for invalid CSS content property values)

twilco committed:
AX: CSS content alt text is not announced by VoiceOver
https://bugs.webkit.org/show_bug.cgi?id=292653
rdar://150830550

Reviewed by Joshua Hoffman.

This commit incorporates usage of alt text from CSS `content` into the various existing places we get alt text, e.g.
from the alt attribute.

* LayoutTests/accessibility/css-content-alt-text-expected.txt: Added.
* LayoutTests/accessibility/css-content-alt-text.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
(WebCore::AccessibilityImageMapLink::description const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::usesAltForTextComputation const):
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::description const):
(WebCore::AccessibilityNodeObject::usesAltTagForTextComputation const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::altTextFromAttributeOrStyle const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::objectInclusionFromAltText):
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::createFor):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::updateAltText):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::altFromContent const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: https://commits.webkit.org/297848@main
